pkg/scm/git/git_test.go

```diff
@@ -10,28 +10,80 @@ import (
 	"testing"
 
 	"github.com/openshift/source-to-image/pkg/api"
+	"github.com/openshift/source-to-image/pkg/errors"
 	"github.com/openshift/source-to-image/pkg/test"
 )
 
+// Creates a git directory with one unlikely but possible commit hash
 func createLocalGitDirectory(t *testing.T) string {
-	dir, err := ioutil.TempDir(os.TempDir(), "s2i-test")
+	dir, err := ioutil.TempDir(os.TempDir(), "gitdir-s2i-test")
 	if err != nil {
 		t.Error(err)
 	}
-	os.Mkdir(filepath.Join(dir, ".git"), 0600)
+	os.MkdirAll(filepath.Join(dir, ".git/refs/heads"), 0777)
+	os.MkdirAll(filepath.Join(dir, ".git/refs/remotes"), 0777)
+	os.MkdirAll(filepath.Join(dir, ".git/branches"), 0777)
+	os.MkdirAll(filepath.Join(dir, ".git/objects/fo"), 0777)
+	os.Create(filepath.Join(dir, ".git/objects/fo") + "12345678901234567890123456789012345678") // 40 character SHA-1 hash
 	return dir
 }
 
-func TestIsLocalGitRepository(t *testing.T) {
+func createEmptyLocalGitDirectory(t *testing.T) string {
+	dir, err := ioutil.TempDir(os.TempDir(), "gitdir-s2i-test")
+	if err != nil {
+		t.Error(err)
+	}
+	os.MkdirAll(filepath.Join(dir, ".git/refs/heads"), 0777)
+	os.MkdirAll(filepath.Join(dir, ".git/refs/remotes"), 0777)
+	os.MkdirAll(filepath.Join(dir, ".git/branches"), 0777)
+	os.MkdirAll(filepath.Join(dir, ".git/objects"), 0777)
+	return dir
+}
+
+func TestIsValidGitRepository(t *testing.T) {
 	d := createLocalGitDirectory(t)
 	defer os.RemoveAll(d)
-	if isLocalGitRepository(d) == false {
+
+	// We have a .git that is populated
+	// Should return true with no error
+	ok, err := isValidGitRepository(d)
+	if !ok {
 		t.Errorf("The %q directory is git repository", d)
 	}
-	os.RemoveAll(filepath.Join(d, ".git"))
-	if isLocalGitRepository(d) == true {
+
+	if err != nil {
+		t.Errorf("isValidGitRepository returned an unexpected error: %q", err.Error())
+	}
+
+	d = createEmptyLocalGitDirectory(t)
+	defer os.RemoveAll(d)
```

Comment:
That's what I was wondering before :) I looked it up, defer evaluates the arguments when called but delays the actual function until the surrounding function returns.